src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.RequestThrottling;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Aspnetcore.RequestThrottling
+{
+    public class RequestThrottlingMiddleware
+    {
+        private SemaphoreWrapper _semaphore;
+
+        private readonly RequestThrottlingOptions _options;
+        private readonly RequestDelegate _next;
+        private readonly ILogger _logger;
+
+        public RequestThrottlingMiddleware(RequestDelegate next, ILoggerFactory loggerFactory, IOptions<RequestThrottlingOptions> options)
+        {
+            _next = next;
+            _logger = loggerFactory.CreateLogger<RequestThrottlingMiddleware>();
+            _options = options.Value;
+            _semaphore = new SemaphoreWrapper(_options.MaxConcurrentRequests); 
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            await _semaphore.EnterQueue();
+            _logger.LogDebug("Entered Queue");
+
+            await _next(context);
```

Comment:
put the call to next in a try block and LeaveQueue in a finally block. That way you don't leak anything if next throws an Exception. This would be a good test case.

Reply:
Or `EnterQueue` could return an `IDisposable` that leaves the queue when disposed. Then you can use `using`.

Reply:
Having implemented both, I'm going with the Try/Finally because the purpose is more obvious and it does not require an additional class.